windows/flutter_nsd_plugin.cpp

```diff
@@ -0,0 +1,327 @@
+        send(it->second);
+      }
+      if (result == -1) return;
+    }
+    delete this;
+  }
+
+
+  void MdnsRequest::ThreadFunc(MdnsRequest *p, std::string serviceName)
+  {
+    p->runner(serviceName);
+  }
+
+
+  void MdnsRequest::start(std::string serviceName)
+  {
+    keepRunning = TRUE;
+    std::thread myThread(ThreadFunc, this, serviceName);
+    myThread.detach();
+  }
+
+  void MdnsRequest::stop()
+  {
+    keepRunning = false;
+  }
+
+
+class FlutterNsdPlugin : public flutter::Plugin {
+ public:
+  static void RegisterWithRegistrar(flutter::PluginRegistrarWindows *registrar);
+
+  FlutterNsdPlugin(std::shared_ptr<flutter::MethodChannel<flutter::EncodableValue>>);
+
+  virtual ~FlutterNsdPlugin();
+
+ private:
+  // Called when a method is called on this plugin's channel from Dart.
+  void HandleMethodCall(
+      const flutter::MethodCall<flutter::EncodableValue> &method_call,
+      std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result);
+  std::shared_ptr<flutter::MethodChannel<flutter::EncodableValue>> channel;
+
+  MdnsRequest* currentRequest;
+
+};
+
+// static
+void FlutterNsdPlugin::RegisterWithRegistrar(
+    flutter::PluginRegistrarWindows *registrar) {
+  auto channel =
+      std::make_shared<flutter::MethodChannel<flutter::EncodableValue>>(
+          registrar->messenger(), "com.nimroddayan/flutter_nsd",
+          &flutter::StandardMethodCodec::GetInstance());
+
+  auto plugin = std::make_unique<FlutterNsdPlugin>(channel);
+
+  channel->SetMethodCallHandler(
+      [plugin_pointer = plugin.get()](const auto &call, auto result) {
+        plugin_pointer->HandleMethodCall(call, std::move(result));
+      });
+
+  registrar->AddPlugin(std::move(plugin));
+}
+
+
+
+FlutterNsdPlugin::FlutterNsdPlugin(std::shared_ptr<flutter::MethodChannel<flutter::EncodableValue>> _channel)
+
+{
+  channel = _channel;
+  currentRequest = NULL;
+
+  WORD versionWanted = MAKEWORD(1, 1);
+  WSADATA wsaData;
+  if (WSAStartup(versionWanted, &wsaData)) {
+
+    printf("Failed to initialize WinSock\n");
+  }
+}
+
+FlutterNsdPlugin::~FlutterNsdPlugin() {
+  WSACleanup();
+}
+
+
+
+void FlutterNsdPlugin::HandleMethodCall(
+  const flutter::MethodCall<flutter::EncodableValue>& method_call,
+  std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
+  if (method_call.method_name().compare("startDiscovery") == 0) {
+    const auto* arguments = std::get_if<flutter::EncodableMap>(method_call.arguments());
+    auto serviceType = arguments->find(flutter::EncodableValue("serviceType"));
+    std::string type = "";
+    if (serviceType != arguments->end()) {
+      type = std::get<std::string>(serviceType->second);
+    }
+
+    if (currentRequest != NULL) {
+      auto req = currentRequest;
+      currentRequest = NULL; // request will delete itself after the thread exits, we can detach the pointer here
+      req->stop(); // will stop the background thread after the next timeout; returns immediately
```

Comment:
What happens if `start()` is called before the next timeout? Would have have 2 scans ongoing?

Reply:
Yes - it would be very difficult to preempt the running query as it may be stuck in a select() call
It should be OK - the only thing is that you may receive results from both queries until the first one times out

Reply:
Sounds good to me 👍